File: Python/PrintInOrder.py
import threading
from typing import Callable

lock1 = threading.Lock()
lock2 = threading.Lock()
lock1.acquire()
lock2.acquire()

class Foo:
    #Submission
    #https://leetcode.com/submissions/detail/575147407/
    
    def __init__(self):
        pass

    
    # printFirst() outputs "first". Do not change or remove this line.
    def first(self, printFirst: 'Callable[[], None]') -> None:
        printFirst()
        lock1.release()
        
    # printSecond() outputs "second". Do not change or remove this line.
    def second(self, printSecond: 'Callable[[], None]') -> None:
        lock1.acquire()
        printSecond()
        lock2.release()
        

        
            
    # printThird() outputs "third". Do not change or remove this line.
    def third(self, printThird: 'Callable[[], None]') -> None:
        lock2.acquire()
        printThird()

File: Java/Completed/Design/PrintInOrder.java
package Completed.Design;
import java.util.concurrent.Semaphore;

//Submission: https://leetcode.com/problems/print-in-order/submissions/1451098605

class PrintInOrder {
    private Semaphore sema2;
    private Semaphore sema3;

    public PrintInOrder() {
        sema2 = new Semaphore(0);
        sema3 = new Semaphore(0);
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printFirst.run();
        sema2.release();
    }

    public void second(Runnable printSecond) throws InterruptedException {
        sema2.acquire();
        printSecond.run();
        sema3.release();
    }

    public void third(Runnable printThird) throws InterruptedException {
        sema3.acquire();
        printThird.run();
    }
}